sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -140,12 +137,21 @@ private[csv] object CSVInferSchema {
     }
   }
 
+  private def tryParseDate(field: String, options: CSVOptions): DataType = {
+    // This case infers a custom `dateFormat` is set.
+    if ((allCatch opt options.dateFormat.parse(field)).isDefined) {
+      DateType
+    } else {
+      tryParseTimestamp(field, options)
+    }
+  }
+
   private def tryParseTimestamp(field: String, options: CSVOptions): DataType = {
-    // This case infers a custom `dataFormat` is set.
+    // This case infers a custom `timestampFormat` is set.
     if ((allCatch opt options.timestampFormat.parse(field)).isDefined) {
       TimestampType
     } else if ((allCatch opt DateTimeUtils.stringToTime(field)).isDefined) {
-      // We keep this for backwords competibility.
+      // We keep this for backwards compatibility.
       TimestampType
     } else {
       tryParseBoolean(field, options)
```

Comment:
You need to change this line. : ) Otherwise, when we infer the schema, we will not enter `tryParseDate`.